商品取得処理用のスタブAPIの実装

ult-sohtome opened the pull request:
## 変更点

- 商品取得処理用のスタブを用意した
　- 商品データは仮のものを用意した
　- 通信失敗や遅延ケースを想定できるような実装とした
- 画面上にボタンを用意し、ボタン押下でFake Store APIからのデータ取得とスタブからのデータ取得を切り替えられるようにした
- ボタン押下で切り替わるのは商品取得処理用APIのみとした

ult-sohtome committed:
商品取得処理用のスタブを作成

ult-sohtome committed:
ボタン押下でデータ取得を'Fake Store APIから'と'スタブから'を切り替える
・ストアで商品取得処理用リポジトリの状態を管理する

ult-sohtome committed:
翻訳APIのスタブ切替実装漏れのコンポーネント修正
・不要なコメント削除

ult-shimizu commented on src/repositories/StubProductRepository.ts:
```diff
@@ -0,0 +1,37 @@
+  ) {}
+
+  private readonly products: Array<Product> = [
+    { id: 1, title: "Product 1", price: 10, description: "Description 1", category: "book", image: "image1.jpg" },
```
商品画像が表示されないのが気になりました。
商品画像を用意するか、もしくは空にする形でお願いしたいです！

ult-sohtome replied:
商品画像を用意することにしました！
[8a9a07f](https://github.com/ult-sohtome/shopping-site/pull/13/commits/8a9a07ff6761e0b6eb2d9f493658015b6085f37b)

ult-shimizu commented on src/stores/UseProductRepositoryStore.ts:
```diff
@@ -0,0 +1,24 @@
+    isUseStub.value = !isUseStub.value;
+    productRepository.value = isUseStub.value
+      ? new StubProductRepository()
+      : new ApiProductRepository();
```
ファクトリー使ってないことが気になりました。
使用箇所ごとにファクトリー使っていたり使っていなかったりなど、一貫性がないとメンテがしずらいです。
createProductRepository()でこのストアのisUseStubを参照する形がよさそうです！

また、isUseStubとproductRepositoryで別に状態を管理している点が気になりました。
１つの状態を管理し、あとは計算すれば十分と考えています。
・isUseStubがtrueの時、スタブのリポジトリ
・isUseStubがfalseの時、APIのリポジトリ
isUseStubのみstateで管理し、productRepositoryはcomputed()を用意するのがよさそうです。

ult-sohtome replied:
通話にてご教授頂きありがとうございました！
ファクトリーを使用すると例えボタンで切り替えていてもisUseStubがtrueになってしまうため、ファクトリーを使っていませんでした。
しかし、productRepositoryをcomputed()にし、.envファイルの環境変数は初期値としてストアのisUseStubで状態管理を統一することで解決できること理解できました！
以下のコミット内で修正しました。
[6e912a2](https://github.com/ult-sohtome/shopping-site/pull/13/commits/6e912a2ca1b9b751c2165d703fb440c3614d243c)

ult-shimizu commented on src/factories/productRepositoryFactory.ts:
```diff
@@ -0,0 +1,10 @@
+import type { ProductRepositoryInterface } from '@/interfaces/ProductRepositoryInterface';
+
+export function createProductRepository(): ProductRepositoryInterface {
+  const isUseStub: boolean = import.meta.env.VITE_USE_PRODUCT_STUB === 'true';
```
こちら新しく用意してもらった変数の認識です。
.env.example、build-environment.mdファイルにも記述の追加をお願いしたいです！

また、商品取得APIを使うか、スタブ実装を使うかですが、今回画面のボタンからも切り替え可能になっています。
ただし、この環境変数がボタンによって更新されることはなさそうです。
（こちらの環境変数はあくまで初期値の認識）
なので、`VITE_DEFAULT_USE_PRODUCT_STUB`のような名前など、初期値であることがわかるような命名でお願いします！

商品APIとカートが不具合
・商品詳細・カート一覧画面でボタン押下しても切り替わらない
・スタブでカート追加（カート一覧にはスタブのデータが表示）商品一覧で本番に切り替え、同一IDのものをカートに追加すると、一覧には本番の商品が２つ表示
・商品idが20（スタブには存在しないid）でカートに追加すると、カート一覧には表示されないものの、pinaのストアにはデータが存在している

修正方針しては、スタブ切り替えボタンを押下すると「カートの商品がすべて削除されます。よろしいですか？」とダイアログが表示され、OKが押されたら、カートの商品をすべて削除…という形でお願いします！
また、スタブ切り替えボタンは商品一覧のみに表示する形でお願いします！

ult-sohtome replied:
環境変数に関して修正しました。
[dd06018](https://github.com/ult-sohtome/shopping-site/pull/13/commits/dd06018782a427db8c3e67d5a7ce8263785002ce)

> 商品APIとカートが不具合

↑この件を改善するため、以下の通りにしました。[6e912a2](https://github.com/ult-sohtome/shopping-site/pull/13/commits/6e912a2ca1b9b751c2165d703fb440c3614d243c)
- スタブ切り替えボタンは商品一覧のみに表示する形に修正
- ボタン押下時ダイアログを表示し、切り替え実行する場合はカート内商品を削除する仕様に修正

ult-shimizu commented on src/stores/UseProductRepositoryStore.ts:
```diff
@@ -0,0 +1,24 @@
+
+export const useProductRepositoryStore = defineStore('ProductRepository', () => {
+  const isUseStub = ref(import.meta.env.VITE_USE_PRODUCT_STUB === 'true');
+  const productRepository = ref<ProductRepositoryInterface>(createProductRepository());
```
データの状態はあまり持ちたくないので、computed()でお願いします！

ult-shimizu reviewed (changes requested):
@ult-sohtome 
PRありがとうございます！数点コメントしたのでご確認をお願いします！

ult-sohtome committed:
stubの商品画像を表示出来るように修正

ult-sohtome committed:
商品取得APIまたはスタブについて手順書、.env.exampleに追記
環境変数名を初期値であるように命名

ult-sohtome committed:
スタブ切り替えボタンを商品一覧ページにのみ表示する
・ボタン押下でカート内商品が削除されることを確認するように修正
・ストアのproductRepositoryをcomputedに修正
・ストアでfactoryを使用してリポジトリ切り替え実装箇所を統一するように修正
・watchの削除

ult-sohtome committed:
カート商品読み込み中が表示されるように修正

ult-sohtome commented:
@ult-shimizu 
コメントおよび通話にてご相談させて頂きありがとうございました！
修正しましたので、ご確認のほどよろしくお願いします。

加えて、１点質問があります。
スタブ切り替えに関してwatch()が無くても現行動作するので不要かと思い削除しました。
しかし、watchが必要なのは今回のように1ページに1コンポーネントではなかった場合に反応してもらうためだったでしょうか？

ult-shimizu commented on src/factories/productRepositoryFactory.ts:
```diff
@@ -1,10 +1,12 @@
-  return isUseStub
+  const productRepositoryStore = useProductRepositoryStore();
+
+  return productRepositoryStore.isUseStub
```
すみません。通話の中で、引数で渡さなくてもstoreを参照すればよい…とお話したのですが、よくよく考えたら相互依存になってしまっていました…。
お手数おかけしますが、isUseStubは引数で受け取る形とさせてください…！

ult-sohtome replied:
ご指摘ありがとうございます！
相互依存が好ましくない理由は、後に一方のみ変更があった際にもう片方が壊れる可能性があったり関係性が複雑になって管理しづらくなったりすることだと知りました。
[3a60ea0](https://github.com/ult-sohtome/shopping-site/pull/13/commits/3a60ea0abce2bee194545b4dda8684590585384e)
このコミットで引数で受け取るように修正しました。

ult-shimizu reviewed (changes requested):
@ult-sohtome 
修正ありがとうございます！
すみません！１点だけ通話の時に考慮不足があったため、コメントしました。

>加えて、１点質問があります。
スタブ切り替えに関してwatch()が無くても現行動作するので不要かと思い削除しました。
しかし、watchが必要なのは今回のように1ページに1コンポーネントではなかった場合に反応してもらうためだったでしょうか？

→削除してもらって大丈夫です！
（すみません！通話の中でwatch必要かも…とお話したのですが、不要でした…）
watchが必要になってくるケースは詳細やカート一覧にボタンがある状態です。
詳細やカート一覧のボタンを削除してもらったので、この画面でwatchする必要がなくなりました（=mountedのタイミングのみ気にすれば十分）

あとはstoreの状態と表示状態を常に同期させておきたい…となったらwatchを使うべきですが、そこまでやる必要はないと考えています。
・現状通常の動作でstoreの値と乖離することはできない
・スタブなので、そこまで作りこむ必要はない

ult-sohtome committed:
相互依存を避けるため引数でisUseStubを受け取るように修正
・watchのimport不要のため削除

ult-sohtome committed:
ダイアログ表示のコンポーネントを作成
・confirm()表示をコンポーネントの使用に修正

ult-sohtome commented:
@ult-shimizu 
コメントありがとうございました！
https://github.com/ult-sohtome/shopping-site/pull/13#discussion_r2108633877
ご指摘いただいた箇所について修正しました。

また、confirm()で表示していたダイアログを、コンポーネントで表示するように修正しました。
[f6ee61d](https://github.com/ult-sohtome/shopping-site/pull/13/commits/f6ee61df1aab040c11eb5be9b1a636ca6c77194d)

以上２点のご確認をお願いいたします。

ult-shimizu reviewed (approved):
@ult-sohtome 
修正ありがとうございます！LGTMです！
